### ISSUE TITLE
[unstable option] unstable_features

### ISSUE THREAD

**Issue by scampi** [rust-lang/rustfmt#3387]
Tracking issue for unstable_features

**Comment by Stargateur** [comment]
So to use `unstable_features` that allow to opt in unstable feature, itself need to be in already unstable channel because it itself not stable.

**Comment by thomaseizinger** [comment]
According to [this issue](https://github.com/rust-lang/rustfmt/issues/3247) it behaves the way it is supposed to behave.

However, I'd still argue that it should behave differently.

Our usecase is to compile our code against stable Rust but we like to use unstable formatting features. Currently, this forces us install two versions of Rust in CI which
- a) causes configuration complexity in our CI setup
- b) extends the time our CI builds take by having to download two toolchains
- c) causes complexity in our builds scripts because we have to call `cargo fmt` with a `+nightly-XXXX` option.

**Comment by topecongiro** [comment]
On 2.0, we will remove this and enable all unstable features on the nightly channel by default. On the stable, every unstable feature will be disabled.

**Comment by thomaseizinger** [comment]
Just to clarify:
There is going to be no way of using unstable formatting features on the stable release channel? :(
So we always have to install two versions of Rust in CI, just to check formatting and then compile with stable?

**Comment by karyon** [comment]
This will need to be re-decided for 1.x

**Comment by Veetaha** [comment]
> So we always have to install two versions of Rust in CI, just to check formatting and then compile with stable?

I totally agree on this one. It's really annoying, and it's the thing across entire rust toolchain. Why do we need to explicitly install the nightly version of the toolchain to use an unstable feature? Why is it not enough to pass `--unstable-features` flag to "sign a contract that I'm using unstable". This makes it annoying to use unstable features and provide feedback for them so much that one would rather give up on the idea and wait until the feature becomes stable.

With `rustc` though there is at least a workaround of using `RUSTC_BOOTSTRAP`, but with rustfmt you have to explicitly download and pin a specific nightly version

**Comment by Veetaha** [comment]
For anyone else who lands here: **YOU CAN USE UNSTABLE FEATURES ON STABLE TOOLCHAN**, but you have to specify them via the CLI parameters like this:

```bash
cargo fmt -- --config unstable_features=true --config error_on_unformatted=true
```

For the reasoning why you can use them only via CLI and can't use them via a config file see https://github.com/rust-lang/rustfmt/issues/5511#issuecomment-1867600019

**Comment by joshka** [comment]
> For anyone else who lands here: **YOU CAN USE UNSTABLE FEATURES ON STABLE TOOLCHAN**, but you have to specify them via the CLI parameters like this:
> 
> ```shell
> cargo fmt -- --config unstable_features=true --config error_on_unformatted=true
> ```
> 
> For the reasoning why you can use them only via CLI and can't use them via a config file see [#5511 (comment)](https://github.com/rust-lang/rustfmt/issues/5511#issuecomment-1867600019)

I was confused about this and thought this meant that I could enable just the unstable_features via cli and that would cause the rustfmt.toml config file values to just work, but you have to enable every config option that you'd like to use. This seems like an unfortunate and non-intuitive way of making this work.

**Comment by utkarshgupta137** [comment]
If you look at the code, what unstable_features does is allow you to unstable features of the rustfmt command, not unstable options. There are 3 rustfmt arguments that you can't use without unstable_features: skip-children, error-on-unformatted, file-lines. You can only use these 3 options if you pass --unstable-features on the CLI or include it in rustfmt.toml.
Source: https://github.com/rust-lang/rustfmt/blob/main/src/bin/main.rs#L562-L597
The source of the confusion is this comment at the top of the configuration docs:

> Each configuration option is either stable or unstable. Stable options can always be used, while unstable options are only available on a nightly toolchain and must be opted into. To enable unstable options, set unstable_features = true in rustfmt.toml or pass --unstable-features to rustfmt.

I think the last sentence should be removed or edited to reflect the reality.
